improve error message when there is an invalid package name

For a package `foo`, I was getting the following error:

```
[foo.0]     Synchronizing with /Users/ashish/code/foo
[ERROR] At ~/.opam/4.01.0/packages.dev/foo/opam
[ERROR] opam file for foo.0 contains errors, ignoring
```

I eventually figured out the problem was that I wrote "future.lwt" in the depends field when I should have written "future" (i.e. I was mixing up ocamlfind package names with opam package names). Unfortunately, the error message gives no indication that this was the problem.

--- a/src/client/opamClient.ml
+++ b/src/client/opamClient.ml
@@ -1518,16 +1518,15 @@ module API = struct
       OpamGlobals.msg "\n";
       let empty_opam = OpamState.has_empty_opam t nv in
       let needs_reinstall2 =
-        if empty_opam then
-          OpamState.add_pinned_overlay ~template:true t name;
         if edit || empty_opam then
-          OpamPinCommand.edit t name
+          try OpamPinCommand.edit t name
+          with Not_found ->
+            (OpamGlobals.error "No valid metadata available.";
+             ignore (unpin name);
+             OpamGlobals.exit 1)
         else None
       in
-      let skip_action =
-        (* Unedited, empty opam *)
-        needs_reinstall2 = None && empty_opam in
-      if action && not skip_action then
+      if action then
         let t = OpamState.load_state "pin-reinstall-2" in
         if not (OpamPackage.has_name t.installed name) ||
            needs_reinstall <> None ||

--- a/src/client/opamPinCommand.ml
+++ b/src/client/opamPinCommand.ml
@@ -42,29 +42,35 @@ let edit t name =
     with Not_found -> None
   in
   let file = OpamPath.Switch.Overlay.opam t.root t.switch name in
-  if not (OpamFilename.exists file) then
-    OpamState.add_pinned_overlay ~template:true t name;
-  let backup =
-    OpamFilename.OP.(OpamPath.backup_dir t.root //
-                     (OpamPackage.Name.to_string name ^ "-opam.bak")) in
-  OpamFilename.copy ~src:file ~dst:backup;
-  let orig_opam = OpamFile.OPAM.read file in
+  let temp_file = OpamPath.Switch.Overlay.tmp_opam t.root t.switch name in
+  let orig_opam =
+    try Some (OpamFile.OPAM.read file) with e -> OpamMisc.fatal e; None
+  in
+  let empty_opam =
+    match orig_opam with
+    | None -> true
+    | Some opam ->
+      try opam = OpamFile.OPAM.create
+            (OpamPackage.create name (OpamFile.OPAM.version opam))
+      with OpamSystem.Internal_error _ -> true
+  in
+  if empty_opam then
+    if not (OpamFilename.exists temp_file)
+    then OpamState.add_pinned_overlay ~template:true t name;
+  if not (OpamFilename.exists temp_file) then
+    OpamFilename.copy ~src:file ~dst:temp_file;
   let rec edit () =
     try
       ignore @@ Sys.command
         (Printf.sprintf "%s %s"
-           (Lazy.force OpamGlobals.editor) (OpamFilename.to_string file));
-      let opam = OpamFile.OPAM.read file in
-      let opam =
-        if OpamFile.OPAM.name_opt opam = None then
-          OpamFile.OPAM.with_name opam name
-        else opam in
+           (Lazy.force OpamGlobals.editor) (OpamFilename.to_string temp_file));
+      let opam = OpamFile.OPAM.read temp_file in
       if OpamFile.OPAM.name_opt opam <> Some name then
-        (OpamGlobals.error "Inconsistent \"name\" field, it must be %s"
+        (OpamGlobals.error
+           "Inconsistent or missing \"name\" field, it must be %S"
            (OpamPackage.Name.to_string name);
          failwith "bad name");
       match pin, OpamFile.OPAM.version_opt opam with
-      | Version vpin, None -> OpamFile.OPAM.with_version opam vpin
       | Version vpin, Some v when v <> vpin ->
         OpamGlobals.error
           "Bad \"version: %S\" field: the package is version-pinned to %s"
@@ -74,37 +80,36 @@ let edit t name =
       | _, None ->
         OpamGlobals.error "The \"version\" field is mandatory";
         failwith "missing field"
-      | _ -> opam
+      | _ -> Some opam
     with e ->
-      (try OpamMisc.fatal e with e ->
-        OpamFilename.move ~src:backup ~dst:file;
-        raise e);
+      OpamMisc.fatal e;
       if OpamGlobals.confirm "Errors in %s, retry editing ?"
           (OpamFilename.to_string file)
       then edit ()
-      else (OpamFilename.move ~src:backup ~dst:file;
-            OpamFile.OPAM.read file)
+      else None
   in
-  let opam = edit () in
-  OpamGlobals.msg "You can edit this file again with \"opam pin edit %s\"\n"
-    (OpamPackage.Name.to_string name);
-  OpamFilename.remove backup;
-  if opam = orig_opam then None else
-  let () = match pin with
-    | Local dir ->
-      let src_opam =
-        OpamFilename.OP.(
-          if OpamFilename.exists_dir (dir / "opam") then dir / "opam" // "opam"
-          else dir // "opam")
-       in
-       if OpamGlobals.confirm "Save the new opam file back to %S ?"
-           (OpamFilename.to_string src_opam) then
-         OpamFilename.copy ~src:file ~dst:src_opam
-    | Version _ | Git _ | Darcs _ | Hg _ -> ()
-  in
-  match installed_nv with
-  | None -> None
-  | Some nv -> Some (OpamPackage.version nv = OpamFile.OPAM.version opam)
+  match edit () with
+  | None -> if empty_opam then (prerr_endline "EO"; raise Not_found) else (prerr_endline "NEO"; None)
+  | Some new_opam ->
+    OpamFilename.move ~src:temp_file ~dst:file;
+    OpamGlobals.msg "You can edit this file again with \"opam pin edit %s\"\n"
+      (OpamPackage.Name.to_string name);
+    if Some new_opam = orig_opam then None else
+    let () = match pin with
+      | Local dir ->
+        let src_opam =
+          OpamFilename.OP.(
+            if OpamFilename.exists_dir (dir / "opam") then dir / "opam" // "opam"
+            else dir // "opam")
+        in
+        if OpamGlobals.confirm "Save the new opam file back to %S ?"
+            (OpamFilename.to_string src_opam) then
+          OpamFilename.copy ~src:file ~dst:src_opam
+      | Version _ | Git _ | Darcs _ | Hg _ -> ()
+    in
+    match installed_nv with
+    | None -> None
+    | Some nv -> Some (OpamPackage.version nv = OpamFile.OPAM.version new_opam)
 
 let update_set set old cur save =
   if OpamPackage.Set.mem old set then
@@ -151,7 +156,10 @@ let pin name pin_option =
           "%s is already pinned to %s."
           (OpamPackage.Name.to_string name)
           (string_of_pin_option current);
-      if OpamGlobals.confirm "Proceed ?" then no_changes
+      if OpamGlobals.confirm "Proceed ?" then
+        (OpamFilename.remove
+           (OpamPath.Switch.Overlay.tmp_opam t.root t.switch name);
+         no_changes)
       else OpamGlobals.exit 0
     with Not_found -> false
   in

--- a/src/client/opamPinCommand.mli
+++ b/src/client/opamPinCommand.mli
@@ -23,7 +23,9 @@ open OpamTypes
 val pin: name -> pin_option -> bool option
 
 (** Let the user edit a pinned package's opam file.
-    Returns [Some is_same_version] if the package should be rebuilt *)
+    Returns [Some is_same_version] if the package should be rebuilt.
+    raises [Not_found] if no valid opam file is available and the user didn't
+    succeed in producing one. *)
 val edit: OpamState.state -> name -> bool option
 
 (** Unpin a package. Returns true if the package should be rebuilt *)

--- a/src/client/opamState.ml
+++ b/src/client/opamState.ml
@@ -480,7 +480,9 @@ let remove_metadata t packages =
 (* Returns [opam, descr_file, files_dir]. We don't consider [url] since
    this is for pinned packages. if [root], don't look for a subdir [opam]
    to find [files] and [descr]. *)
-let local_opam ?(root=false) ?(version_override=true) nv dir =
+let local_opam
+    ?(root=false) ?(version_override=true) ?copy_invalid_to
+    nv dir =
   let has_dir d = if OpamFilename.exists_dir d then Some d else None in
   let has_file f = if OpamFilename.exists f then Some f else None in
   let opam_dir, descr, files_dir =
@@ -504,9 +506,14 @@ let local_opam ?(root=false) ?(version_override=true) nv dir =
         else Some opam
       with e ->
         OpamMisc.fatal e;
-        OpamGlobals.error "opam file for %s contains errors, ignoring"
-          (OpamPackage.to_string nv);
-        None
+        match copy_invalid_to with
+        | Some dst ->
+          OpamGlobals.warning
+            "Errors in opam file from %s source, \
+             ignored (fix with 'opam pin edit')"
+            (OpamPackage.to_string nv);
+          OpamFilename.copy ~src:local_opam ~dst; None
+        | None -> None
   in
   opam, descr, files_dir
 
@@ -2465,7 +2472,11 @@ let update_dev_package t nv =
     (* Do the update *)
     let result = fetch () in
     let new_meta = (* New version from the source *)
-      hash_meta @@ local_opam ~version_override:false nv srcdir
+      hash_meta @@
+      local_opam
+        ~version_override:false
+        ~copy_invalid_to:(OpamPath.Switch.Overlay.tmp_opam t.root t.switch name)
+        nv srcdir
     in
     let user_meta, old_meta, repo_meta =
       if OpamFilename.exists (srcdir // "opam") then
@@ -2510,6 +2521,8 @@ let update_dev_package t nv =
         (OpamGlobals.msg "Installing new package description for %s from %s\n"
            (OpamPackage.to_string nv)
            (Filename.concat (string_of_address remote_url) "opam");
+         OpamFilename.remove
+           (OpamPath.Switch.Overlay.tmp_opam t.root t.switch name);
          install_meta srcdir user_meta new_meta)
       else if
         OpamGlobals.msg

--- a/src/client/opamSwitchCommand.ml
+++ b/src/client/opamSwitchCommand.ml
@@ -281,7 +281,7 @@ let switch_cont ~quiet ~warning switch =
     OpamState.check_base_packages t
 
 let switch ~quiet ~warning switch =
-  (snd (switch_cont ~quiet ~warning switch)) () 
+  (snd (switch_cont ~quiet ~warning switch)) ()
 
 (* Remove from [set] all the packages whose names appear in
    [filter]. *)

--- a/src/core/opamPath.ml
+++ b/src/core/opamPath.ml
@@ -168,6 +168,8 @@ module Switch = struct
 
     let opam t a n = package t a n // "opam"
 
+    let tmp_opam t a n = package t a n // "opam_"
+
     let url t a n = package t a n // "url"
 
     let descr t a n = package t a n // "descr"

--- a/src/core/opamPath.mli
+++ b/src/core/opamPath.mli
@@ -234,6 +234,10 @@ module Switch: sig
         $opam/$switch/cache/$name.$version/opam} *)
     val opam: t -> switch -> name -> filename
 
+    (** OPAM temp overlay (for user editing): {i
+        $opam/$switch/cache/$name.$version/opam_} *)
+    val tmp_opam: t -> switch -> name -> filename
+
     (** URL overlay: {i
         $opam/$switch/overlay/$name.$version/url} *)
     val url: t -> switch -> name -> filename

